lib/ssh/src/ssh_agent.erl

```diff
@@ -0,0 +1,93 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2019. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+%% Reference: https://tools.ietf.org/html/draft-miller-ssh-agent-02
+
+-module(ssh_agent).
+
+-include("ssh.hrl").
+-include("ssh_agent.hrl").
+
+-export([send/1, pack/1, unpack/1, encode/1, decode/1]).
+
+%% Agent communication
+
+send(Request) ->
+    SocketPath = os:getenv("SSH_AUTH_SOCK"),
+
+    ConnectOpts = [binary, {packet, 0}, {active, false}],
+    {ok, Socket} = gen_tcp:connect({local, SocketPath}, 0, ConnectOpts),
+
+    BinRequest = ssh_agent:pack(ssh_agent:encode(Request)),
+    ok = gen_tcp:send(Socket, BinRequest),
+
+    Timeout = 1000, % TODO: Make this a parameter? What is a sensible default value?
```

Comment:
It is always nice to have timeouts configurable.  1000 seems like a sensible value given that the ssh-agent is local.
There is a mechanism for passing options to for  example ssh_file_with_agent:user_key/2 with a call like
`ssh_file_with_agent:user_key(Alg, [{timeout,1234}])`
which origins from
`ssh:connect(Host, Port, [{key_cb, {ssh_file_with_agent,[{timeout,1234}]}}])`

See ssh_file:user_key/2 for an example of options passing to ssh_file, the man page [Module:user_key-2](http://erlang.org/doc/man/ssh_client_key_api.html#Module:user_key-2) and the connect option [key_cb](http://erlang.org/doc/man/ssh.html#type-key_cb_common_option)

Reply:
I've update `ssh_agent:send/1,2` to take an additional `Timeout` argument defaulting to `infinity` which appears to be a common default value for timeouts in OTP (probably because it's somewhat "neutral" and doesn't make any assumption for how long something should take).

All the current usages of this function (in `ssh_auth` and the new `ssh_file_with_agent` modules) will explicitly provide a timeout value of `1000` if the callback module's `timeout` option is not specified.

Here's an updated usage example:

```erlang
HostName = "127.0.0.1",
Opts = [{key_cb, {ssh_file_with_agent, [{timeout, 500}]}}],
{ok, Conn} = ssh:connect(HostName, 22, Opts),
% …
```


https://gist.github.com/pmeinhardt/a39494d43f6acf606fc7249a3fe2cb50